### DISCUSSION TITLE
RPLanguage contrib obfuscate empty

### DISCUSSION THREAD

**Discussion by CharlesMccown** [evennia/evennia#2856] [Evennia Questions] [answered]
Using RPLanguage and I have several languages implemented. 

Most of them work well, but one language loosely based on gaelic frequently obfuscates words to the empty string.

Anyone have ideas to ensure that all words get translated in an obfuscated message or know if it's possible to simply have a bad grammar that could result in words not being translated?

If there's a problem with my grammar, phonemes or vowels that seems like the best solution.
```
phonemes = "a á e é i o ó u ù b bh " \
               "c ch d dh f fh g gh l m " \
               "mh n p ph r s sh t th"
vowels = "aáeéioóuù"
grammar = "v cv vc cvc vcv cvv vvc cvvc cvcc ccvc " \
          "cvvcv cvvcc cvccv ccvcv ccvcc cvvcvv " \
          "cvvcvc cvvccv cvcvvc cvcvcc cvvcvvcc " \
          "cvvccvv cvvcvvc cvccvvc ccvcvvc ccvvcvv " \
          "cvvccvvv cvvccvvc cvccvccvv ccvcvccv"
```

If there's nothing obviously wrong with the language definition above, then my  naive first thought was to split the message on whitespace and ensure there are no missing words, and if there are to try obfuscating the original message again. 

The concerns with this approach are:
- there is probably something I'm missing to force all words to translate
- could this result in a race condition or other issues with repeatedly trying and failing to generate all words in a sentence

I'd be very grateful for any input or ideas. It just feels like there's an easy and obvious solution that I'm missing.

Thank you.

**Comment by Griatch** [comment, marked as answer]
I see no obvious issue with your config. There are two main and normal reasons for a (pseudo-)translation to be empty: 
-  You have set a large word-length variation and the original word is short. In this case, 0 is a valid length of the 'translated' word. Allowing to drop some words makes sense - another language would not always use exactly the same number of words as English to say something. It adds to the obfuscation too. 
- The other reason for an empty word is because your word-length variation caused the system to look for a word of a certain length and there is no phoneme + grammar combination that can be found that matches this. Again, a 0-length word is allowed, because not all words would be expected to have an equivalent translation.

**Reply by Zireael07** [reply]
Can we have a boolean to control whether 0 length 'words' are allowed? Sometimes it makes sense, sometimes it does not...

**Reply by Griatch** [reply]
It would make little sense to limit this. Either set a small word-length variation or make a manual translation for words that you want a fixed translation for.

**Reply by CharlesMccown** [reply]
Thanks Griatch.

I'm creating manual translations for most of the "important" words for my languages. I'll include the list below because it includes the top 100 commonly used words in English, so maybe it will be useful to others.

I'm going to continue debugging this to try and figure out what's going on with my words. The variation for the language I'm having trouble with is 1 and I have added grammar's for length up to 10, but the word "something" never translates so I now have a good test case.

In case anyone else would like the template of manual translations that I'm currently using, feel free to borrow the word list below. I've tried to remove all words specific to my world, but may have missed a few.
<details>
  <summary>Manual Translations</summary>

```
manual_translations = {
        "I": "",
        "I'll": "",
        "I'm": "",
        "OK": "",
        "a": "",
        "about": "",
        "above": "",
        "air": "",
        "all": "",
        "and": "",
        "ant": "",
        "ants": "",
        "are": "",
        "as": "",
        "at": "",
        "back": "",
        "bad": "",
        "bang": "",
        "battle": "",
        "battles": "",
        "be": "",
        "because": "",
        "bee": "",
        "been": "",
        "bees": "",
        "beetle": "",
        "beetles": "",
        "below": "",
        "beneath": "",
        "big": "",
        "boom": "",
        "bug": "",
        "bugs": "",
        "but": "",
        "buzz": "",
        "can": "",
        "can't": "",
        "cave": "",
        "cavern": "",
        "caverns": "",
        "cities": "",
        "city": "",
        "come": "",
        "could": "",
        "damselflies": "",
        "damselfly": "",
        "dark": "",
        "darkness": "",
        "death": "",
        "deep": "",
        "defend": "",
        "demon": "",
        "demons": "",
        "depth": "",
        "depths": "",
        "devil": "",
        "devils": "",
        "did": "",
        "didn't": "",
        "do": "",
        "don't": "",
        "down": "",
        "dragon": "",
        "dragonflies": "",
        "dragonfly": "",
        "dragons": "",
        "dwarf": "",
        "dwarve": "",
        "dwarven": "",
        "dwarves": "",
        "earth": "",
        "elf": "",
        "elfs": "",
        "elven": "",
        "elves": "",
        "evil one": "",
        "evil": "",
        "faithful": "",
        "father": "",
        "fathers": "",
        "fire": "",
        "fires": "",
        "flies": "",
        "fly": "",
        "for": "",
        "forbidden one": "",
        "forge": "",
        "forgot": "",
        "forgotten": "",
        "from": "",
        "frost": "",
        "get": "",
        "glory": "",
        "gnat": "",
        "gnats": "",
        "gnome": "",
        "gnomes": "",
        "go": "",
        "going": "",
        "good one": "",
        "good": "",
        "got": "",
        "guard": "",
        "guards": "",
        "had": "",
        "hail": "",
        "have": "",
        "he": "",
        "he's": "",
        "hello": "",
        "her": "",
        "here": "",
        "hero": "",
        "heroes": "",
        "hey": "",
        "high": "",
        "hill": "",
        "hillock": "",
        "hillocks": "",
        "hills": "",
        "him": "",
        "his": "",
        "holy": "",
        "hornet": "",
        "hornets": "",
        "how": "",
        "if": "",
        "in": "",
        "iron": "",
        "is": "",
        "it": "",
        "it's": "",
        "just": "",
        "kingdom": "",
        "kingdoms": "",
        "know": "",
        "lair": "",
        "lairs": "",
        "legend": "",
        "legends": "",
        "like": "",
        "look": "",
        "man": "",
        "me": "",
        "mean": "",
        "men": "",
        "misery": "",
        "mithril": "",
        "mount": "",
        "mountain": "",
        "mountains": "",
        "mounts": "",
        "my": "",
        "no": "",
        "not": "",
        "now": "",
        "of": "",
        "oh": "",
        "okay": "",
        "on": "",
        "one": "",
        "or": "",
        "out": "",
        "pain": "",
        "peak": "",
        "peaks": "",
        "place": "",
        "pop": "",
        "really": "",
        "realm": "",
        "realms": "",
        "right": "",
        "river": "",
        "rivers": "",
        "say": "",
        "see": "",
        "she": "",
        "sleep": "",
        "small": "",
        "so": "",
        "some": "",
        "something": "",
        "state": "",
        "tell": "",
        "that": "",
        "that's": "",
        "the": "",
        "then": "",
        "there": "",
        "they": "",
        "think": "",
        "this": "",
        "time": "",
        "to": "",
        "town": "",
        "towns": "",
        "under": "",
        "unspeakable": "",
        "up": "",
        "valley": "",
        "valleys": "",
        "village": "",
        "villages": "",
        "villain": "",
        "villains": "",
        "want": "",
        "was": "",
        "water": "",
        "we": "",
        "well": "",
        "were": "",
        "what": "",
        "when": "",
        "who": "",
        "why": "",
        "will": "",
        "with": "",
        "woman": "",
        "women": "",
        "word": "",
        "words": "",
        "world": "",
        "worlds": "",
        "would": "",
        "yeah": "",
        "yes": "",
        "you": "",
        "you're": "",
        "your": ""
    }
```
 <details>

**Reply by CharlesMccown** [reply]
So, I found the issue and it's a bit embarrassing.

**I didn't have a vv phoneme, but I had vv combinations in the grammar.**

Adding ae and áe to the phoneme list seems to have fixed everything.

![image](https://user-images.githubusercontent.com/16038455/189778244-44d29655-113f-4017-ac42-938a6ba45f3c.png)

Thanks again!

**Reply by Griatch** [reply]
Aha, I didn't notice that in your grammar either. It's an easy mistake to make. Glad you found it :+1:
